### ISSUE TITLE
Can this work on OS X?

### ISSUE THREAD

**Issue by breschke** [TiesdeKok/ipystata#6]
How do I (can I) register a stata instance on OS X 10.11?

**Comment by breschke** [comment]
Please make OS dependency plainer in doc.

**Comment by TiesdeKok** [comment]
You are right, the Windows dependency is not clear enough currently. I plan to add a slimmed down version (which does not rely on Stata automation) which will work on Linux and OS X, I hope to soon find some time to work on it!

**Comment by breschke** [comment]
Excellent. Very interested in both Linux and OS X-compatible versions.

**Comment by TiesdeKok** [comment]
I have written a draft version which should work on Linux and OS X, I have tested it on Linux (Debian 7.9) and the main things appear to work (the --changewd argument does not seem to work yet). 

Could you try to run it to see if it works on OS X? Just run the code in this notebook: https://dl.dropboxusercontent.com/u/1265025/ipystata_batchmode.ipynb

**Comment by cschwem2er** [comment]
Hey this is great news! I only have ubuntu, but this should work as well right? Could you explain how this batch mode works?

**Comment by breschke** [comment]
The following block of code returns an error:
`iPyStata.config_stata('/Applications/Stata/StataSE.app/Contents/MacOS/StataSE')`

``` ---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-5-0aa5496aca94> in <module>()
----> 1 iPyStata.config_stata('/Applications/Stata/StataSE.app/Contents/MacOS/StataSE')

AttributeError: iPyStata instance has no attribute 'config_stata'
```

This is testing on OS X 10.11.2

**Comment by TiesdeKok** [comment]
@breschke My bad, I forgot to include a function. See this updated notebook: https://dl.dropboxusercontent.com/u/1265025/ipystata_batchmode_v1.ipynb

@methodds 
Yes, this should also work on Ubuntu. The batch mode is a less-sophisticated way to interact with Stata compared to using Stata Automation. In a nutshell, my code takes the cell-code and generates a temporary .do file which is then executed via the command line (see for example http://www.stata.com/support/faqs/unix/batch-mode/). The feature-set is more limited (no persistence across cells, no macro retrieval back to Python, only one session per notebook) but at least it works consistently.

**Comment by breschke** [comment]
@TiesdeKok Great, it seems to be working properly. Will likely try out on Linux next. Been hoping for this kind of support for Stata for quite some time. Thanks!

**Comment by TiesdeKok** [comment]
@breschke Good! 

I have worked a bit on the graph and working directory functionality. I (kind of) got it to work. Apparently the Unix console version of Stata does not allow exporting of graphs to an image format directly because it does not render it in a display window (see http://www.stata.com/statalist/archive/2011-07/msg00536.html). This is quite inconvenient because the only remaining options are either .EPS or .PDF which I both cannot directly display as output to the Jupyter Notebook. PDF files can be displayed using a HTML iframe but the problem is that the underlying Tornado web server (which runs the Jupyter Notebook) cannot open local files from directories other than those in the starting directory (safety reasons). 

After fiddling around with it for a couple of hours I figured out a "best-I-can-do" workaround: 
1. When ipystata is imported it saves the current working directory ("hoping" it is the starting directory)
2. Stata saves the graph as a .pdf file to this supposed starting directory 
3. The PDF is opened and displayed via an iframe. 

This approach has two downsides:
1. If ipystata is imported after the python working directory is changed it might break the functionality (because the pdf is attempted to be loaded from "above" the starting directory).
2. Temporary pdf files are saved in the notebook folder until ipystata is unloaded (it will try to remove the file when ipystata is closed)

I hope this approach also works on OS X, could you try it out for me? See this notebook:
https://dl.dropboxusercontent.com/u/1265025/ipystata_batchmode_v2.ipynb

**Comment by TiesdeKok** [comment]
@breschke Have you had the chance to test my latest notebook (v2) using OS X? Unfortunately I cannot test it myself because I do not have an OS X instance with Stata. Once I know it works I will complete the code and include it into IPyStata.

**Comment by jaschwer** [comment]
For me it does not work (Mac, Yosemite).
I run your code and Stata opens and I see that it loaded the dataset, but I don't get any results.

I had to make one change in your code, I hope that does not cause the trouble.
I inserted 
'iPyStata.config_stata('/Applications/Stata/StataSE.app/Contents/MacOS/statase')'
and
'stata = '/Applications/Stata/StataSE.app/Contents/MacOS/statase''

No Results means that when I should get (None, None) after running 
'stata = '/Applications/Stata/StataSE.app/Contents/MacOS/statase'
cmd = [stata, "do", 'code.do']
p = subprocess.Popen(cmd)
p.communicate()'
I don't get anything in the notebook. Can you help me with that?

**Comment by TiesdeKok** [comment]
@jaschwer Thanks for letting me know! Does the remainder of the code work if you comment out that cell (the code below)? Does it show the graph / tables as output?

```
stata = '/home/user/stata14/stata'
cmd = [stata, "do", 'code.do']
p = subprocess.Popen(cmd)
p.communicate()
```

It appears that Stata is found correctly given that you mention Stata opens up, which makes me wonder whether the subprocess module behaves differently on OS X by muting the output if it is empty.

**Comment by jaschwer** [comment]
When I closed stata manuelly, I actually got the output (None, None)

So I comment out 
`p = subprocess.Popen(cmd)
p.communicate()`

Then I run the different codes and what happens is that for each code-segment I do, I get the table, regression and so on within stata, but not in my notebook. If I however close Stata manuelly, I get the tables afterwards in notebook. But I need to do that procedure for every segment of code, i.e.
closing after
'%%stata -o car_df
sysuse auto.dta'
and after 
'%%stata -d car_df
tabulate headroom foreign'

and so on.

But should I really see the tables within stata or not just directly in my notebook?

**Comment by TiesdeKok** [comment]
You should see all the outputs appear directly in the notebook, Stata is supposed to operate in the background. 

I suspect this behavior is caused by the `.communicate()` function, this is the part that "idles" my code until Stata is ready. Only after that it starts to process the output and display it in the notebook, this would explain why it does show an output after you close Stata manually.

Let me see what is causing this!

**Edit:**
The problem is quite simple I think, the Stata batch mode is different between OS X and Linux.

Just like in Windows an additional argument is required to make Stata close the program after it is done, after it is closed `.communicate()` releases it and an output is shown in the notebook.

Can you try to run this? 

```
stata = '/home/user/stata14/stata'
cmd = [stata, "-e" ,"do", 'code.do']
p = subprocess.Popen(cmd)
p.communicate()
```

**Comment by jaschwer** [comment]
I get the (None. None) now directly (after waiting for some seconds), but I still have to close stata to get see e.g. "(1978 Automobile Data)" in the output. And it still runs first the command in stata itself. Sorry.

**Comment by TiesdeKok** [comment]
That is good, can you try to replace the following in the first cell with all the code:

```
cmd = [stata, "do", 'code.do']
```

With:

```
cmd = [stata, "-e", "do", 'code.do']
```

It is near the bottom just above:

```
    try:
        p = subprocess.Popen(cmd)
        p.communicate()
    except:
        return "Failed to open Stata"
```

**Comment by jaschwer** [comment]
Great! It works! Thank you very much :+1:

**Comment by TiesdeKok** [comment]
Awesome, thanks for trying it out! I will modify my code to deal with this difference between Linux and OS X and then decide on the best way to include it into an actual release. :smile:

**Comment by TiesdeKok** [comment]
Thanks all for the testing efforts, see the new 0.3.0 release for Mac OS X and Linux support! If you encounter any issues with that build please don't hesitate to open a new issue.

**Comment by jaschwer** [comment]
It might be a more general problem, but i run the command
'!pip install ipystata --upgrade --force-reinstall' 

in my notebook and I get that everything is installed correctly. Then I want to import 'ipystata' I get the following error:

ImportError                               Traceback (most recent call last)
<ipython-input-12-b40be68f6678> in <module>()
----> 1 import ipystata
      2 from ipystata.config import config_stata

ImportError: No module named ipystata

Do you have an idead what goes wrong or is it something more mac specific?

**Comment by TiesdeKok** [comment]
You need to run that particular command in the appropriate command line and environment. I suspect that the error is triggered by it installing in a different environment than the one you use to run the notebook. If you use Anaconda I would recommend to use the "Anaconda Command Prompt" and run the pip command there (you should not run the pip command from within the notebook).

**Comment by jaschwer** [comment]
I am very new to all of this. Can you suggest me a link where it is explained to fix that problem?

**Comment by TiesdeKok** [comment]
Of course. You basically install Python packages using your Mac OS X terminal, `pip` is a package manager that makes it easy to install Python packages. So all you have to do (assuming you only have one Python installation) is:
1. Open up your terminal
2. Run: `pip install ipystata`

Anaconda is a distribution that makes it easy because it comes with a lot of cool packages pre-installed, see: https://docs.continuum.io/anaconda/install

If you still run into problems then you can send me a message in our Gitter room: https://gitter.im/TiesdeKok/ipystata#
